File: frontend/src/services/engineerTicketService.js
import { api } from './apiClient'

/**
 * Engineer routes (/engineer/tickets): the caller's own queue. Engineers only; any ticket
 * not assigned to the caller is a 404. Loaders take (params, {signal}) for useApiData.
 */

/**
 * The tickets assigned to the caller, P1 first, then the longest-waiting. Empty values
 * are ignored.
 * @param {{view?: 'active'|'closed', status?: string, priority?: string,
 *   building_id?: number|string, q?: string}} [filters]
 * @param {{signal?: AbortSignal}} [options]
 */
export function listMyQueue(filters = {}, { signal } = {}) {
  return api.get('/engineer/tickets', { params: filters, signal })
}

/**
 * One of the caller's tickets, with priority and the requester's contact details.
 * @param {{ticketId: number|string}} params
 * @param {{signal?: AbortSignal}} [options]
 */
export function getAssignedTicket({ ticketId }, { signal } = {}) {
  return api.get(`/engineer/tickets/${Number(ticketId)}`, { signal })
}

/** Every status one of the caller's tickets has been in, oldest first. */
export function listAssignedTicketHistory({ ticketId }, { signal } = {}) {
  return api.get(`/engineer/tickets/${Number(ticketId)}/history`, { signal })
}

/** Every note on one of the caller's tickets, oldest first. */
export function listAssignedTicketNotes({ ticketId }, { signal } = {}) {
  return api.get(`/engineer/tickets/${Number(ticketId)}/notes`, { signal })
}

/** Add a note to one of the caller's tickets. Rejects with 409 if it's closed. */
export function addAssignedTicketNote(ticketId, noteText) {
  return api.post(`/engineer/tickets/${Number(ticketId)}/notes`, { note_text: noteText.trim() })
}

/**
 * Move one of the caller's tickets along the workflow. Resolves with the updated ticket.
 * A blank reason is left out (it's required for blocked and resolved; the API says so
 * with a 422). Rejects with 409 for a move the workflow doesn't allow.
 * @param {number|string} ticketId
 * @param {'in_progress'|'blocked'|'resolved'} status
 * @param {string} [reason]
 */
export function changeTicketStatus(ticketId, status, reason = '') {
  const body = { status }
  if (reason.trim()) body.reason = reason.trim()
  return api.post(`/engineer/tickets/${Number(ticketId)}/status`, body)
}

File: backend/core/services/engineer_ticket_service.py
"""An engineer's own queue: the tickets currently assigned to them, with priority and the
requester's contact details, plus their notes and status history.

Any other ticket (unassigned, someone else's, or reassigned away) is "not found", so an
engineer can't learn about tickets that aren't theirs. Status changes come in E2.
"""

from typing import Any

import psycopg

import db
from errors import ConflictError, NotFoundError
from repositories import ticket_repository
from schemas import EngineerTicketFilters


def _get_assigned_ticket(conn: psycopg.Connection, user_id: int, ticket_id: int) -> dict[str, Any]:
    """Return the ticket if it's assigned to this engineer, else raise NotFoundError."""
    ticket = ticket_repository.get_admin_detail(conn, ticket_id)
    if ticket is None or ticket["assigned_to_user_id"] != user_id:
        raise NotFoundError("Ticket not found")
    return ticket


def list_my_queue(user_id: int, filters: EngineerTicketFilters) -> list[dict[str, Any]]:
    """Return the engineer's tickets matching the filters, P1 first then longest-waiting."""
    closed = None if filters.view is None else filters.view == "closed"
    with db.transaction() as conn:
        return ticket_repository.list_all(
            conn,
            status=filters.status,
            priority=filters.priority,
            urgency=None,
            category=None,
            building_id=filters.building_id,
            assigned=None,
            assigned_to=user_id,
            escalated=None,
            closed=closed,
            search=filters.q or None,
        )


def get_ticket(user_id: int, ticket_id: int) -> dict[str, Any]:
    """Return full details of one of the engineer's tickets."""
    with db.transaction() as conn:
        return _get_assigned_ticket(conn, user_id, ticket_id)


def list_notes(user_id: int, ticket_id: int) -> list[dict[str, Any]]:
    """Return every note on one of the engineer's tickets, oldest first."""
    with db.transaction() as conn:
        _get_assigned_ticket(conn, user_id, ticket_id)
        return ticket_repository.list_notes(conn, ticket_id)


def list_status_history(user_id: int, ticket_id: int) -> list[dict[str, Any]]:
    """Return every status one of the engineer's tickets has been in, oldest first."""
    with db.transaction() as conn:
        _get_assigned_ticket(conn, user_id, ticket_id)
        return ticket_repository.list_status_history(conn, ticket_id)


def add_note(user_id: int, ticket_id: int, note_text: str) -> dict[str, Any]:
    """Add a note to one of the engineer's unclosed tickets and bump its updated_at.

    The ticket row is locked, so it can't be reassigned or closed while the note is added.
    """
    with db.transaction() as conn:
        ticket_repository.lock(conn, ticket_id)
        ticket = _get_assigned_ticket(conn, user_id, ticket_id)
        if ticket["status"] == "closed":
            raise ConflictError("Closed tickets can't take new notes")
        note = ticket_repository.insert_note(conn, ticket_id, user_id, note_text)
        ticket_repository.touch(conn, ticket_id)
        return note
